### PR TITLE
Take a single root node in range_search

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
The unsafe code can be justified within range_search, as it makes sure to not
overlap the returned references, but from the callers perspective it's an
entirely safe algorithm and there's no need for the caller to know about the
duplication.

cc @ssomers
r? @Amanieu

### CODE REVIEW

**Comment by ssomers** [comment]
By the way, I was wondering why it wasn't marked unsafe. Not only could you monkey around by passing in different root, it still returns two mutable handles into the same tree and you can invalidate one handle by changing the tree through the other. Not that I think sprinkling more unsafe words helps readability.

What would help readability, I think, is moving the function out of the fat map.rs file into search.rs. Although if you pass a `RangeFull` as range, it acts like the first_leaf_edge/last_leaf_edge functions in navigate.rs.

File: src/liballoc/collections/btree/map.rs [hunk 1]
```diff
@@ -2064,8 +2059,10 @@ where
         _ => {}
     };
 
-    let mut min_node = root1;
-    let mut max_node = root2;
+    // We duplicate the root NodeRef here -- we will never access it in a way
+    // that overlaps references obtained from the root.
```
Comment (ssomers) [inline src/liballoc/collections/btree/map.rs:2063]: I don't understand that sentence (but I don't have to). In an emptied tree, range_search will return two references to the same and only edge. Is that no overlap?

**Comment by Mark-Simulacrum** [comment]
Yeah, I think the justification here may not be perfect -- but I do think this is the right move. In particular, it seems like the risk of "well you could invalidate the tree" is sort of always true -- we detach the handles from the root node pretty frequently -- and I wouldn't want to add a bunch of unsafe everywhere, for sure not until we have the unsafe-in-unsafe-fns lint.

**Comment by Amanieu** [comment]
I personally think that the entirety of the internal btree code should just be unsafe, so I don't feel too strongly about this.

@bors r+ rollup

**Pull request merged**